### PR TITLE
Spaces various fixes and polish

### LINKED ISSUE #3291: DMs shouldn't light up spaces
As per web change
https://github.com/vector-im/element-web/issues/17044

### PR DESCRIPTION
**Pull request by BillCarsonFr**
- Fix spaces add rooms via servers: re-use the code made for permalink factory
- Home Space now shows all rooms (lab option to orphan only)
image

Support more depth in space pannel, + smaller items for sub spaces
image

- Show badge on notification section header

image

Fixes #3291

### CODE REVIEW

File: vector/src/main/java/im/vector/app/features/home/room/list/RoomListViewModel.kt [hunk 1]
```diff
@@ -69,12 +69,12 @@ class RoomListViewModel @Inject constructor(
...
          * If current space is null, will return orphan rooms only
          */
-        NORMAL,
+        ORPHANS_IF_SPACE_NULL,
         /**
          * Special case when we don't want to discriminate rooms when current space is null.
          * In this case return all.
          */
-        NOT_IF_ALL,
+        ALL_IF_SPACE_NULL,
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/list/RoomListViewModel.kt:77]: ❤️

File: vector/src/main/java/im/vector/app/features/home/room/list/SpaceRoomListSectionBuilder.kt [hunk 2]
```diff
@@ -87,11 +89,13 @@ class SpaceRoomListSectionBuilder(
...
-                        activeSpaceAwareQueries,
-                        R.string.invitations_header,
-                        true,
-                        RoomListViewModel.SpaceFilterStrategy.NORMAL
+                        sections = sections,
+                        activeSpaceUpdaters = activeSpaceAwareQueries,
+                        nameRes = R.string.invitations_header,
+                        notifyOfLocalEcho = true,
+                        spaceFilterStrategy = RoomListViewModel.SpaceFilterStrategy.ORPHANS_IF_SPACE_NULL.takeIf { onlyOrphansInHome }
+                                ?: RoomListViewModel.SpaceFilterStrategy.ALL_IF_SPACE_NULL,
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/list/SpaceRoomListSectionBuilder.kt:97]: I would prefer a more readable `if else` block here

File: vector/src/main/java/im/vector/app/features/home/room/list/SpaceRoomListSectionBuilder.kt [hunk 3]
```diff
@@ -102,7 +106,8 @@ class SpaceRoomListSectionBuilder(
                         activeSpaceAwareQueries,
                         R.string.bottom_action_rooms,
                         false,
-                        RoomListViewModel.SpaceFilterStrategy.NORMAL
+                        RoomListViewModel.SpaceFilterStrategy.ORPHANS_IF_SPACE_NULL.takeIf { onlyOrphansInHome }
+                                ?: RoomListViewModel.SpaceFilterStrategy.ALL_IF_SPACE_NULL
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/list/SpaceRoomListSectionBuilder.kt:110]: same remark

File: vector/src/main/java/im/vector/app/features/home/room/list/SpaceRoomListSectionBuilder.kt [hunk 4]
```diff
@@ -152,7 +159,8 @@ class SpaceRoomListSectionBuilder(
                 activeSpaceAwareQueries,
                 R.string.bottom_action_rooms,
                 false,
-                RoomListViewModel.SpaceFilterStrategy.NORMAL
+                RoomListViewModel.SpaceFilterStrategy.ORPHANS_IF_SPACE_NULL.takeIf { onlyOrphansInHome }
+                        ?: RoomListViewModel.SpaceFilterStrategy.ALL_IF_SPACE_NULL
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/list/SpaceRoomListSectionBuilder.kt:163]: same

File: vector/src/main/java/im/vector/app/features/home/room/list/SpaceRoomListSectionBuilder.kt [hunk 5]
```diff
@@ -164,7 +172,8 @@ class SpaceRoomListSectionBuilder(
                 activeSpaceAwareQueries,
                 R.string.low_priority_header,
                 false,
-                RoomListViewModel.SpaceFilterStrategy.NORMAL
+                RoomListViewModel.SpaceFilterStrategy.ORPHANS_IF_SPACE_NULL.takeIf { onlyOrphansInHome }
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/list/SpaceRoomListSectionBuilder.kt:175]: same. Create a private `fun`?

File: vector/src/main/java/im/vector/app/features/spaces/SpaceSummaryController.kt [hunk 6]
```diff
@@ -132,13 +134,13 @@ class SpaceSummaryController @Inject constructor(
         }
 
         rootSpaces
-                ?.sortedBy { it.displayName }
+                ?.sortedBy { it.roomId }
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/spaces/SpaceSummaryController.kt:137]: sorting by `roomId` is strange...

File: vector/src/main/java/im/vector/app/features/spaces/SubSpaceSummaryItem.kt [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+@EpoxyModelClass(layout = R.layout.item_sub_space)
+abstract class SubSpaceSummaryItem : VectorEpoxyModel<SubSpaceSummaryItem.Holder>() {
+
+    @EpoxyAttribute(EpoxyAttribute.Option.DoNotHash)  lateinit var avatarRenderer: AvatarRenderer
+    @EpoxyAttribute lateinit var matrixItem: MatrixItem
+    @EpoxyAttribute var selected: Boolean = false
+    @EpoxyAttribute(EpoxyAttribute.Option.DoNotHash) var listener: (() -> Unit)? = null
+    @EpoxyAttribute(EpoxyAttribute.Option.DoNotHash)  var onMore: (() -> Unit)? = null
+    @EpoxyAttribute(EpoxyAttribute.Option.DoNotHash)  var toggleExpand: (() -> Unit)? = null
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/spaces/SubSpaceSummaryItem.kt:45]: (minor) bad formatting

File: vector/src/main/java/im/vector/app/features/spaces/SubSpaceSummaryItem.kt [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+                            if (expanded) R.drawable.ic_expand_less else R.drawable.ic_expand_more
+                    )
+            )
+            holder.collapseIndicator.setOnClickListener(
+                    DebouncedClickListener({ _ ->
+                        toggleExpand?.invoke()
+                    })
+            )
+        } else {
+            holder.collapseIndicator.isGone = true
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/spaces/SubSpaceSummaryItem.kt:80]: remove from `if else` block and write `holder.collapseIndicator.isVisible = hasChildren` ?

File: vector/src/main/res/layout/item_sub_space.xml [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        android:layout_width="26dp"
+        android:layout_height="26dp"
+        android:layout_gravity="center"
+        android:layout_marginStart="@dimen/layout_horizontal_margin"
+        android:contentDescription="@string/avatar"
+        android:duplicateParentState="true"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintStart_toEndOf="@id/indent"
+        app:layout_constraintTop_toTopOf="parent"
+        tools:src="@tools:sample/avatars" />
```
Comment (bmarty) [inline vector/src/main/res/layout/item_sub_space.xml:34]: Reminder to myself to update that once PR about samples is merged

**Review by bmarty — COMMENTED** [review]
Some remarks, did not test it on a device yet

File: vector/src/main/java/im/vector/app/core/ui/list/GenericPillItem.kt [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+    override fun bind(holder: Holder) {
+        super.bind(holder)
+
+        holder.textView.setTextOrHide(text)
+        holder.textView.typeface = style.toTypeFace()
+        holder.textView.textSize = style.toTextSize()
+        holder.textView.gravity = if (centered) Gravity.CENTER_HORIZONTAL else Gravity.START
+
+        imageRes?.let { holder.imageView.setImageResource(it) }
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/core/ui/list/GenericPillItem.kt:68]: updated

File: vector/src/main/java/im/vector/app/features/spaces/SpaceSummaryController.kt (hunk 6 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/spaces/SpaceSummaryController.kt:137]: yeah, for now it's just to be consistent across client, pending MSC

File: vector/src/main/java/im/vector/app/features/spaces/SubSpaceSummaryItem.kt (hunk 8 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/spaces/SubSpaceSummaryItem.kt:80]: updated

**Comment by bmarty** [comment]
Lint is not happy:

image

Can you have a look @BillCarsonFr ?

File: vector/src/main/res/layout/item_sub_space.xml (hunk 9 above)
Comment (bmarty) [reply vector/src/main/res/layout/item_sub_space.xml:34]: Done in dfc8860ade5b944bc3464b787a52246aced29e1e

**Pull request merged**